File: src/schemas/discovery.ts
/**
 * Read-only architecture-source discovery.
 *
 * Locates immutable architecture sources for repository-contract verification.
 * No product runtime behavior exists in Phase 1.
 */

import { readFileSync, existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { createHash } from "node:crypto";

const __dirname = dirname(fileURLToPath(import.meta.url));

/** Repository root, resolved from this module's location. */
export const REPO_ROOT = resolve(__dirname, "..", "..");

/** Path to the immutable architecture source directory. */
export const ARCHITECTURE_DIR = resolve(REPO_ROOT, "docs", "architecture");

/** Path to the source manifest. */
export const SOURCE_MANIFEST_PATH = resolve(ARCHITECTURE_DIR, "SOURCE_MANIFEST.json");

/** Path to the working schema mirror directory. */
export const SCHEMAS_DIR = resolve(REPO_ROOT, "schemas");

/** Path to the working example mirror directory. */
export const EXAMPLES_DIR = resolve(REPO_ROOT, "examples");

/** Compute SHA-256 of a file's contents. */
export function sha256File(path: string): string {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

/** Read and parse the source manifest. */
export function readSourceManifest(): Record<string, unknown> {
  if (!existsSync(SOURCE_MANIFEST_PATH)) {
    throw new Error(`Source manifest not found: ${SOURCE_MANIFEST_PATH}`);
  }
  return JSON.parse(readFileSync(SOURCE_MANIFEST_PATH, "utf-8")) as Record<string, unknown>;
}

/** List required architecture Markdown basenames. */
export function requiredArchitectureBasenames(): string[] {
  return [
    "EXPFLOW_WORKFLOW_CURRENT.md",
    "EXPFLOW_CONCEPT_PAPER_V2_3.md",
    "EXPFLOW_IMPLEMENTATION_SPEC_V2_3.md",
    "EXPFLOW_PROTOCOL_SPEC_V2_3.md",
    "EXPFLOW_PROJECT_SNAPSHOT_V2_3.md",
    "Note-On-Architecture.md",
    "V2_3_REVIEW_RESOLUTION.md",
    "V2_3_ARCHITECTURE_DELTA.md",
    "RELATED_WORK.md",
  ];
}

File: python/expflow_hooks/discovery.py
"""Read-only discovery of architecture schema-source directory.

Phase 1 scaffold only. No product runtime behavior exists.

Architecture sources are repository-contract inputs and are not packaged into
the Python wheel. Installed wheels can import the package and report the
contract version, while architecture discovery is explicitly available only
from an editable repository checkout.
"""

from pathlib import Path

_REPO_ROOT = Path(__file__).resolve().parent.parent.parent
ARCHITECTURE_DIR = _REPO_ROOT / "docs" / "architecture"
SCHEMAS_DIR = ARCHITECTURE_DIR / "schemas"
SOURCE_MANIFEST_PATH = ARCHITECTURE_DIR / "SOURCE_MANIFEST.json"
ARCHITECTURE_DISCOVERY_MODE = "editable-repository"


def architecture_sources_available() -> bool:
    """Return whether repository architecture sources are available."""
    return (
        ARCHITECTURE_DIR.is_dir()
        and SCHEMAS_DIR.is_dir()
        and SOURCE_MANIFEST_PATH.is_file()
    )


def require_repository_architecture() -> None:
    """Require architecture sources from an editable repository checkout."""
    if architecture_sources_available():
        return

    missing = [
        path
        for path in (ARCHITECTURE_DIR, SCHEMAS_DIR, SOURCE_MANIFEST_PATH)
        if not path.exists()
    ]
    missing_text = ", ".join(str(path) for path in missing)
    raise FileNotFoundError(
        "Expflow Phase 1 Python architecture discovery is available only from "
        "an editable repository checkout; installed wheels do not package "
        f"docs/architecture. Missing: {missing_text}"
    )


__all__ = [
    "ARCHITECTURE_DIR",
    "ARCHITECTURE_DISCOVERY_MODE",
    "SCHEMAS_DIR",
    "SOURCE_MANIFEST_PATH",
    "architecture_sources_available",
    "require_repository_architecture",
]
